### PR TITLE
New submission permission middleware

### PR DESCRIPTION
**Pull request by loneil**
# Description
New middleware to check SUBMISSION permissions on route. This is to prep for all our submission edit workflow work and fixes the bug in https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-1986.

Include logging (kube probe) change too.

## Types of changes

Bug fix (non-breaking change which fixes an issue) 
New feature (non-breaking change which adds functionality) 

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-160

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-160

File: app/src/forms/auth/middleware/userAccess.js [hunk 1]
```diff
@@ -74,6 +76,60 @@ const hasFormPermissions = (permissions) => {
...
+      // No submission provided to this route that secures based on form... that's a problem!
+      return next(new Problem(401, { detail: 'Submission Id not found on request.' }));
+    }
+
+    // Get the submission results so we know what form this submission is for
+    const submissionForm = await service.getSubmissionForm(submissionId);
+    const isDeleted = submissionForm.submission.deleted;
+
+    // Deleted submissions are inaccessible
+    if (isDeleted) {
```
Comment (jujaga) [inline app/src/forms/auth/middleware/userAccess.js:97]: Doesn't seem like isDeleted is used anywhere else. Could streamline this into
```js
if (submissionForm.submission.deleted) {
```

File: app/src/forms/auth/middleware/userAccess.js [hunk 2]
```diff
@@ -74,6 +76,60 @@ const hasFormPermissions = (permissions) => {
...
+    // Does the user have permissions for this submission due to their FORM permissions
+    if (req.currentUser) {
+      let formFromCurrentUser = req.currentUser.forms.find(f => f.formId === submissionForm.form.id);
+      if (formFromCurrentUser) {
+
+        // Do they have the submission permissions being requested on this FORM
+        const intersection = permissions.filter(p => {
+          return formFromCurrentUser.permissions.includes(p);
+        });
+        if (intersection.length == permissions.length) {
```
Comment (jujaga) [inline app/src/forms/auth/middleware/userAccess.js:116]: ===

File: app/src/forms/auth/service.js [hunk 3]
```diff
@@ -212,8 +214,41 @@ const service = {
...
+  // Submission Data
+  // -------------------------------------------------------------------------------------------------------------
+  checkSubmissionPermission: async (currentUser, submissionId, permissions) => {
+    if (currentUser.public) return false;
+    return FormSubmissionUserPermissions.query()
+      .modify('filterSubmissionId', submissionId)
+      .modify('filterUserId', currentUser.id)
+      .modify('filterByPermissions', permissions)
+      .first();
+  },
```
Comment (jujaga) [inline app/src/forms/auth/service.js:230]: Is this query missing a `throwIfNotFound()` in the chain?

File: app/src/forms/form/service.js [hunk 4]
```diff
@@ -281,26 +281,6 @@ const service = {
     }
   },
 
-  updateSubmission: async (formSubmissionId, data, currentUser) => {
```
Comment (jujaga) [inline app/src/forms/form/service.js:284]: I'm not sure if we should be removing this function yet? We'll likely be using this in the future when you end up editing a submission (submitter and reviewer back and forth interaction). It'd just be missing the logic to append to a log table (doesn't exist yet obviously).

File: app/src/forms/submission/routes.js [hunk 5]
```diff
@@ -2,34 +2,36 @@ const routes = require('express').Router();
 
 const controller = require('./controller');
 const currentUser = require('../auth/middleware/userAccess').currentUser;
+const hasSubmissionPermissions = require('../auth/middleware/userAccess').hasSubmissionPermissions;
```
Comment (jujaga) [inline app/src/forms/submission/routes.js:5]: If we change the userAccess middleware file down the line to module.export an object with those functions as attributes, we can then do something nicer like the following:
```js
const { currentUser, hasSubmissionPermissions } = require('../auth/middleware/userAccess');
```

File: app/src/forms/submission/service.js [hunk 6]
```diff
@@ -38,60 +35,15 @@ const service = {
...
-
-    // check against the submission level permissions assigned to the user...
-    const submissionPermission = await checkSubmissionPermission();
-    if (submissionPermission) return result;
-
-    // no access to this submission...
-
-    throw new Problem(401, 'You do not have access to this submission.');
+  read: async (formSubmissionId) => {
+    return await service._fetchSubmissionData(formSubmissionId);
```
Comment (jujaga) [inline app/src/forms/submission/service.js:39]: If we're expecting a promise from _fetchSubmissionData and read function is already async, we could make this function a synchronous passthrough instead and avoid unnecessary async/await tackons:
```js
  read: (formSubmissionId) => service._fetchSubmissionData(formSubmissionId),
```

File: app/src/forms/auth/middleware/userAccess.js (hunk 1 above)
Comment (loneil) [reply app/src/forms/auth/middleware/userAccess.js:97]: Yeah true

File: app/src/forms/form/service.js (hunk 4 above)
Comment (loneil) [reply app/src/forms/form/service.js:284]: Yeah it's unused in the form service but definitely will be used in the submission service (duplicated there so removed this one)

File: app/src/forms/auth/service.js (hunk 3 above)
Comment (loneil) [reply app/src/forms/auth/service.js:230]: No this one is supposed to return which permission intersections you have for the supplied id, user, permission list. So the array of permissions if so, or falsey if nothing.

**Review by jujaga — APPROVED** [review]

**Review by TimCsaky — APPROVED** [review]
this is great. seems like it will make applying permissions a lot easier

**Pull request merged**